src/stores/CartStore.ts

```diff
@@ -0,0 +1,70 @@
+import { makeAutoObservable, observable, runInAction } from "mobx";
+import memoize  from "lodash.memoize";
+import { Product } from "../interfaces/ProductAndCartItemInterface";
+import { CartItem } from "../interfaces/ProductAndCartItemInterface";
+class CartStore {
+    totalItems = observable.box(0);
+    items = observable.array<CartItem>([]);
+
+    setItems(items: CartItem[]) : void
+    {
+        runInAction(() =>{
+            this.items.replace(items);
+        })
+    }
+    addItem(product: Product): void {
+        const existingItem = this.items.find(item => item.id === product.id);
+        let updatedItems;
+        if (existingItem) {
+            updatedItems = this.items.map(item =>
+                item.id === product.id ? { ...item, quantity: item.quantity + 1 } : item
+            );
+        } else {
+            updatedItems = [...this.items, { ...product, quantity: 1 }];
+        }
+        this.setItems(updatedItems);
+        console.log("Items after adding:", this.items);
+        this.updateTotalItems();
+    }
+
+    removeItem(productId: number): void {
+        const itemIndex = this.items.findIndex(item => item.id === productId);
+        if (itemIndex > -1) {
+            let updatedItems;
+            const currentItem = this.items[itemIndex];
+            if (currentItem.quantity > 1) {
+                updatedItems = this.items.map(item =>
+                    item.id === productId ? { ...item, quantity: item.quantity - 1 } : item
+                );
+            } else {
+                updatedItems = this.items.filter(item => item.id !== productId);
+            }
+            this.setItems(updatedItems);
+            console.log("Items after removing:", this.items);
+            this.updateTotalItems();
+        }
+    }
+
+    updateTotalItems(): void {
+        const newTotalItems = this.getTotalItems();
+        console.log("Updating total items to:", newTotalItems);
+        this.totalItems.set(newTotalItems);
+    }
+
+    getTotalItems(): number {
+        return this.items.reduce((total, item) => total + item.quantity, 0);
+    }
+
+    get totalPrice(): number {
+        return this.items.reduce((total, item) => total + item.price * item.quantity, 0);
+    }
+
+    get renderedItems() {
+        return this.items.map(item => ({
+            ...item,
+            removeItem: () => this.removeItem(item.id)
+        }));
+    }
+}
+
+export const cartStore = new CartStore();
```

Comment:
search how to add memoize here

Reply:
Memoization is used to optimize performance by storing the results of expensive function calls and returning the cached result when the same inputs occur again. Here, it prevents unnecessary recalculations of total items and total price, especially when the items array hasn't changed.


Added memoize to the getTotalItems  method within the constructor. This method is now memoized to avoid recalculating its value unless the items array changes.

Reply:
but I think in this app memoize function is not necessary

Reply:
question is not answered. should set here memoize

Reply:
here an example of memoization:
const multiplyBy10 = (num) => {
     return num * 10;
}

const memoizedMyltiplyBy10 = () => {
      const cache = {};
       return (num) =>  {
          if (num in cache){
             console.log(cache);
              return cache[num];
          }
         const result = num * 10;
          cache[num] = result;
           return result;
        }
}

Reply:
so as we can see hereif we have a call 10 time to multiply by 10 form example, so the calculation is occuring once and then being saved in the cache so when we want it, we can access it directly from the cache